matrix/common/query_constructor.py

```diff
@@ -0,0 +1,228 @@
+    the project metadata tsvs. These names are different from the table and column
+    names inside the matrix service. This function takes a filter with a field
+    like "project.project_core.project_short_name" and translates it to
+    "project.short_name" so the query will execute correctly.
+    """
+
+    new_filter = dict(filter_)
+    if "field" in filter_:
+        try:
+            column_name = constants.METADATA_FIELD_TO_TABLE_COLUMN[filter_["field"]]
+            table_name = constants.TABLE_COLUMN_TO_TABLE[column_name]
+            fq_name = table_name + '.' + column_name
+        except KeyError:
+            fq_name = filter_["field"]
+
+        new_filter["field"] = fq_name
+    else:
+        try:
+            new_values = [translate_filters(f) for f in filter_["value"]]
+            new_filter["value"] = new_values
+        except KeyError:
+            pass
+
+    return new_filter
+
+
+def translate_fields(fields):
+    """Translate field list from external metadata field names to internal
+    redshift names.
+    """
+
+    new_fields = []
+    for field in fields:
+        try:
+            column_name = constants.METADATA_FIELD_TO_TABLE_COLUMN[field]
+            table_name = constants.TABLE_COLUMN_TO_TABLE[column_name]
+            fq_name = table_name + '.' + column_name
+        except KeyError:
+            fq_name = field
+        new_fields.append(fq_name)
+    return new_fields
+
+
+def create_matrix_request_queries(filter_, fields, feature):
+    """Based on values from the matrix request, create an appropriate
+    set of redshift queries to serve the request.
+    """
+
+    cell_where_clause = filter_to_where(translate_filters(filter_))
+    feature_where_clause = feature_to_where(feature)
+
+    expression_query = EXPRESSION_QUERY_TEMPLATE.format(
+        feature_where_clause=feature_where_clause,
+        cell_where_clause=cell_where_clause)
+
+    cell_query = CELL_QUERY_TEMPLATE.format(
+        fields=', '.join(translate_fields(fields)),
+        cell_where_clause=cell_where_clause)
+
+    feature_query = FEATURE_QUERY_TEMPLATE.format(feature_where_clause=feature_where_clause)
+
+    return {
+        "expression_query": expression_query,
+        "cell_query": cell_query,
+        "feature_query": feature_query
+    }
+
+
+def feature_to_where(matrix_feature):
+    """Build the WHERE clause for the features."""
+
+    if matrix_feature == "gene":
+        return "feature.isgene"
+    elif matrix_feature == "transcript":
+        return "(NOT feature.isgene)"
+    else:
+        raise MalformedMatrixFeature(f"Unknown feature type {matrix_feature}")
+
+
+def filter_to_where(matrix_filter):
+    """Build a WHERE clause for the matrix request SQL query out of the matrix
+    filter object.
+    """
+
+    try:
+        op = matrix_filter["op"]
+    except Exception as exc:
+        raise MalformedMatrixFilter("Could not retrieve filter op") from exc
```

Comment:
For errors that fail the request, log an error in Dynamo via `request_tracker.log_error` to propagate the failure to the user on GET. These errors could be captured in the driver.

Reply:
Okay, added an `log_error` where this is called in the driver.